### PR TITLE
[ODS-4522] Docker Compose file for shared instance deployment orchestration

### PR DESCRIPTION
**Pull request by smundheinnive**
1- Updated docker-compose-prod.yml with changing the Build names 

2-Updated the nginx.conf in Web-Gateway folder 

3-Updated Web-Ods Dockerfile

4-Added a new setting file webapi.appsettings.production.json

### CODE REVIEW

**Review by gmcelhanon — CHANGES_REQUESTED** [review]
For me, the `ed-fi-ods-docker_gateway_1` container seems to be crashing and restarting with the following entries in the log:
```
/docker-entrypoint.sh: /docker-entrypoint.d/ is not empty, will attempt to perform configuration
/docker-entrypoint.sh: Looking for shell scripts in /docker-entrypoint.d/
/docker-entrypoint.sh: Launching /docker-entrypoint.d/10-listen-on-ipv6-by-default.sh
10-listen-on-ipv6-by-default.sh: Getting the checksum of /etc/nginx/conf.d/default.conf
10-listen-on-ipv6-by-default.sh: Enabled listen on IPv6 in /etc/nginx/conf.d/default.conf
/docker-entrypoint.sh: Launching /docker-entrypoint.d/20-envsubst-on-templates.sh
/docker-entrypoint.sh: Configuration complete; ready for start up
nginx: [emerg] unexpected end of file, expecting "}" in /etc/nginx/nginx.conf:64
```

Also, on the database server on container `ed-fi-ods-docker_db-ods_1`, port 5002, I couldn't see any databases. Looks like (from the log) there was a problem with locating the minimal template file.
```
2020-11-19 00:53:37.161 UTC [30] LOG:  database system is ready to accept connections
 done
server started

/usr/local/bin/docker-entrypoint.sh: running /docker-entrypoint-initdb.d/init-database.sh
CREATE DATABASE
CREATE DATABASE
GRANT
GRANT
/app/EdFi_Ods_Minimal_Template.sql: No such file or directory
2020-11-19 00:53:38.770 UTC [1] LOG:  listening on IPv4 address "0.0.0.0", port 5432
2020-11-19 00:53:38.770 UTC [1] LOG:  listening on IPv6 address "::", port 5432
2020-11-19 00:53:38.800 UTC [1] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
2020-11-19 00:53:38.810 UTC [19] LOG:  database system was interrupted; last known up at 2020-11-19 00:53:37 UTC
2020-11-19 00:53:38.840 UTC [19] LOG:  database system was not properly shut down; automatic recovery in progress
2020-11-19 00:53:38.841 UTC [19] LOG:  redo starts at 0/165C3B8
2020-11-19 00:53:38.841 UTC [19] LOG:  invalid record length at 0/165CCC8: wanted 24, got 0
2020-11-19 00:53:38.841 UTC [19] LOG:  redo done at 0/165CC80
2020-11-19 00:53:38.841 UTC [19] LOG:  last completed transaction was at log time 2020-11-19 00:53:37.684702+00
2020-11-19 00:53:38.844 UTC [1] LOG:  database system is ready to accept connections
```

The other database server seems to be working as expected, so long as the `EdFi_Admin` database is expected to be completely empty?

File: docker-compose-prod.yml [hunk 1]
```diff
@@ -7,46 +7,46 @@ version: '3.8'
...
       - "5402:5432"
     volumes:
-      - ${PWD}/PSQL-ODS/data:/var/lib/postgresql/data
+      - ${PWD}/DB-ODS/data:/var/lib/postgresql/data
     restart: always
 
-  webapi:
-    build: WebApi
+  api:
+    build: Web-Ods
```
[the remark this answers is not in the document]
Comment (smundheinnive) [reply docker-compose-prod.yml:45]: modified the nginx.conf file and updated the init-database.sh in DB-ODS

**Comment by gmcelhanon** [comment]
I tried running this again, but it still didn't work.

I used the following command:
```
docker-compose --env-file .env -f .\docker-compose-prod.yml up -d
```
I used the following environment file:
```
POSTGRES_USER=postgres1
POSTGRES_PASSWORD=pgpassword1
TAG=latest
restart_policy=always
```
I got the following output:
```
WARNING: The PWD variable is not set. Defaulting to a blank string.
Creating ed-fi-ods-docker_db-ods_1   ... done
Creating ed-fi-ods-docker_db-admin_1 ... done
Creating ed-fi-ods-docker_api_1      ... done
Creating ed-fi-ods-docker_gateway_1  ... done
```

`docker ps` returns the following:
```
CONTAINER ID        IMAGE                       COMMAND                  CREATED             STATUS                        PORTS                    NAMES
df20d26da56d        ed-fi-ods-docker_gateway    "/docker-entrypoint.…"   6 minutes ago       Restarting (1) 1 second ago                            ed-fi-ods-docker_gateway_1
a824ef2d0fdd        ed-fi-ods-docker_api        "dotnet EdFi.Ods.Web…"   6 minutes ago       Up 6 minutes                  80/tcp                   ed-fi-ods-docker_api_1
81cf80cfd075        ed-fi-ods-docker_db-ods     "docker-entrypoint.s…"   6 minutes ago       Up 6 minutes                  0.0.0.0:5402->5432/tcp   ed-fi-ods-docker_db-ods_1
c2e5a51a7aae        ed-fi-ods-docker_db-admin   "docker-entrypoint.s…"   6 minutes ago       Up 6 minutes                  0.0.0.0:5401->5432/tcp   ed-fi-ods-docker_db-admin_1
```
`docker logs ed-fi-ods-docker_gateway_1` returns:
```
/docker-entrypoint.sh: /docker-entrypoint.d/ is not empty, will attempt to perform configuration
/docker-entrypoint.sh: Looking for shell scripts in /docker-entrypoint.d/
/docker-entrypoint.sh: Launching /docker-entrypoint.d/10-listen-on-ipv6-by-default.sh
10-listen-on-ipv6-by-default.sh: Getting the checksum of /etc/nginx/conf.d/default.conf
10-listen-on-ipv6-by-default.sh: Enabled listen on IPv6 in /etc/nginx/conf.d/default.conf
/docker-entrypoint.sh: Launching /docker-entrypoint.d/20-envsubst-on-templates.sh
/docker-entrypoint.sh: Configuration complete; ready for start up
nginx: [emerg] unexpected end of file, expecting "}" in /etc/nginx/nginx.conf:64
```

The `ed-fi-ods-docker_db-admin_1` has both the security and admin databases installed, but the EdFi_Admin database is completely empty (is this intentional?).

The `ed-fi-ods-docker_db-ods_1` Postgres container has 2 databases: 

- `EdFi_Ods_Minimal_Template`
- `EdFi_Ods_Populated_Template`

Is that intended for a shared instance deployment?
Also, both database are completely empty (there are no schemas or tables).

**Review by gmcelhanon — CHANGES_REQUESTED** [review]
For me, the `ed-fi-ods-docker_gateway_1` container seems to be crashing and restarting with the following entries in the log:
```
/docker-entrypoint.sh: /docker-entrypoint.d/ is not empty, will attempt to perform configuration
/docker-entrypoint.sh: Looking for shell scripts in /docker-entrypoint.d/
/docker-entrypoint.sh: Launching /docker-entrypoint.d/10-listen-on-ipv6-by-default.sh
10-listen-on-ipv6-by-default.sh: Getting the checksum of /etc/nginx/conf.d/default.conf
10-listen-on-ipv6-by-default.sh: Enabled listen on IPv6 in /etc/nginx/conf.d/default.conf
/docker-entrypoint.sh: Launching /docker-entrypoint.d/20-envsubst-on-templates.sh
/docker-entrypoint.sh: Configuration complete; ready for start up
nginx: [emerg] unexpected end of file, expecting "}" in /etc/nginx/nginx.conf:64
```

Also, on the database server on container `ed-fi-ods-docker_db-ods_1`, port 5002, I couldn't see any databases. Looks like (from the log) there was a problem with locating the minimal template file.
```
2020-11-19 00:53:37.161 UTC [30] LOG:  database system is ready to accept connections
 done
server started

/usr/local/bin/docker-entrypoint.sh: running /docker-entrypoint-initdb.d/init-database.sh
CREATE DATABASE
CREATE DATABASE
GRANT
GRANT
/app/EdFi_Ods_Minimal_Template.sql: No such file or directory
2020-11-19 00:53:38.770 UTC [1] LOG:  listening on IPv4 address "0.0.0.0", port 5432
2020-11-19 00:53:38.770 UTC [1] LOG:  listening on IPv6 address "::", port 5432
2020-11-19 00:53:38.800 UTC [1] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
2020-11-19 00:53:38.810 UTC [19] LOG:  database system was interrupted; last known up at 2020-11-19 00:53:37 UTC
2020-11-19 00:53:38.840 UTC [19] LOG:  database system was not properly shut down; automatic recovery in progress
2020-11-19 00:53:38.841 UTC [19] LOG:  redo starts at 0/165C3B8
2020-11-19 00:53:38.841 UTC [19] LOG:  invalid record length at 0/165CCC8: wanted 24, got 0
2020-11-19 00:53:38.841 UTC [19] LOG:  redo done at 0/165CC80
2020-11-19 00:53:38.841 UTC [19] LOG:  last completed transaction was at log time 2020-11-19 00:53:37.684702+00
2020-11-19 00:53:38.844 UTC [1] LOG:  database system is ready to accept connections
```

The other database server seems to be working as expected, so long as the `EdFi_Admin` database is expected to be completely empty?

**Comment by gmcelhanon** [comment]
@smundheinnive - Please see my updated [review comment](https://github.com/Ed-Fi-Alliance-OSS/Ed-Fi-ODS-Docker/pull/15#pullrequestreview-535008828).

File: compose-shared-instance-env.yml [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+     hostname: nginx
+     volumes:
+        - ./Web-Gateway/nginx-shared.conf:/etc/nginx/nginx.conf
+        - ./Web-Gateway/ssl:/ssl/
+     depends_on:
+        - api
+
+  api:
+    build:
+      context: ./Web-Ods
```
Comment (gmcelhanon) [inline compose-shared-instance-env.yml:54]: @jamessmckay Do you want this artifact (and all usages of it) renamed from `Web-Ods` to `Web-Ods-Api`, as I previously suggested?

Comment (jamessmckay) [reply compose-shared-instance-env.yml:54]: this is the current folder structure
image
So for now yes, if we want to change it, then I suggest we create a ticket.

**Review by gmcelhanon — COMMENTED** [review]
* `POSTGRES_USER` environment variable must be `postgres` or this will not work. ODS database creation fails.
* If `POSTGRES_USER` and `POSTGRES_PASSWORD` environment variables aren't explicitly set, should we stop the process? (Are there well-known defaults for this situation?)
* SSL creation steps in ReadMe.md didn't work for me. My wsl console window wasn't staying open on launch and so I tried using git bash and received the following error at the end of certificate generation:
```
Generating a RSA private key
....++++
...........++++
writing new private key to 'ssl/server.key'
-----
name is expected to be in the format /type0=value0/type1=value1/type2=... where characters may be escaped by \. This name is not in that format: 'C:/Program Files/Git/CN=localhost'
problems making Certificate Request
```
I ended up using openssl with the following command to generate my certificate, which enabled the `ed-fi-gateway` to start up successfully, but I don't know how to verify anything beyond that yet.
```
openssl.exe req -x509 -newkey rsa:4096 -keyout c:\path\to\Ed-Fi-ODS-Docker\Web-Gateway\ssl\server.key -out c:\path\to\Ed-Fi-ODS-Docker\Web-Gateway\server.crt -days 365 -subj /CN=localhost
```

**Review by gmcelhanon — APPROVED** [review]

File: compose-shared-instance-env.yml [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    ports:
+      - "5402:5432"
+    volumes:
+      -  pg-db-ods:/var/lib/postgresql/data
+    restart: always
+    container_name: ed-fi-db-ods
+
+  db-admin:
+    build:
+      context: ./DB-Sandbox
```
Comment (vimayya) [inline compose-shared-instance-env.yml:25]: @gmcelhanon , @jamessmckay  Should this be ./DB-Admin ?

Comment (vimayya) [reply compose-shared-instance-env.yml:25]: Created ODS-4702

**Pull request merged**